blockindex/contractstaking/cache.go

```diff
@@ -71,27 +76,35 @@ func (s *contractStakingCache) CandidateVotes(candidate address.Address) *big.In
 		bt := s.mustGetBucketType(bi.TypeIndex)
 		votes.Add(votes, bt.Amount)
 	}
-	return votes
+	return votes, nil
 }
 
-func (s *contractStakingCache) Buckets() []*Bucket {
+func (s *contractStakingCache) Buckets(height uint64) ([]*Bucket, error) {
 	s.mutex.RLock()
 	defer s.mutex.RUnlock()
 
+	if err := s.validateHeight(height); err != nil {
+		return nil, err
+	}
+
 	vbs := []*Bucket{}
 	for id, bi := range s.bucketInfoMap {
 		bt := s.mustGetBucketType(bi.TypeIndex)
 		vb := assembleBucket(id, bi.clone(), bt, s.contractAddress)
 		vbs = append(vbs, vb)
 	}
-	return vbs
+	return vbs, nil
 }
 
-func (s *contractStakingCache) Bucket(id uint64) (*Bucket, bool) {
+func (s *contractStakingCache) Bucket(id, height uint64) (*Bucket, bool, error) {
 	s.mutex.RLock()
 	defer s.mutex.RUnlock()
 
-	return s.getBucket(id)
+	if err := s.validateHeight(height); err != nil {
+		return nil, false, err
+	}
+	bt, ok := s.getBucket(id)
+	return bt, ok, nil
```

Comment:
`BucketInfo` without height?

Reply:
Currently, it is only necessary to add height restrictions to the indexer's API.